File: expressjs/src/metadata/loader.ts
import type { Queryable } from "../store/postgres.js";
import type { Registry } from "./registry.js";
import type { Entity, Relation } from "./types.js";
import type { Rule, RuleDefinition } from "./rule.js";
import type { StateMachine, StateMachineDefinition } from "./state-machine.js";
import { normalizeDefinition } from "./state-machine.js";
import type { Workflow, WorkflowTrigger, WorkflowStep } from "./workflow.js";
import { normalizeWorkflowSteps } from "./workflow.js";
import type { Permission, PermissionCondition } from "./permission.js";
import type { Webhook, WebhookRetry } from "./webhook.js";

export async function loadAll(
  pool: Queryable,
  registry: Registry,
): Promise<void> {
  const entities = await loadEntities(pool);
  const relations = await loadRelations(pool);
  registry.load(entities, relations);

  const rules = await loadRules(pool);
  registry.loadRules(rules);

  const machines = await loadStateMachines(pool);
  registry.loadStateMachines(machines);

  const workflows = await loadWorkflows(pool);
  registry.loadWorkflows(workflows);

  const permissions = await loadPermissions(pool);
  registry.loadPermissions(permissions);

  const webhooks = await loadWebhooks(pool);
  registry.loadWebhooks(webhooks);

  console.log(
    `Loaded ${entities.length} entities, ${relations.length} relations, ${rules.length} rules, ${machines.length} state machines, ${workflows.length} workflows, ${permissions.length} permissions, ${webhooks.length} webhooks into registry`,
  );
}

export async function reload(
  pool: Queryable,
  registry: Registry,
): Promise<void> {
  return loadAll(pool, registry);
}

async function loadEntities(pool: Queryable): Promise<Entity[]> {
  const result = await pool.query(
    "SELECT name, definition FROM _entities ORDER BY name",
  );
  const entities: Entity[] = [];
  for (const row of result.rows) {
    try {
      const def =
        typeof row.definition === "string"
          ? JSON.parse(row.definition)
          : row.definition;
      entities.push(def as Entity);
    } catch (err) {
      console.warn(`WARN: skipping entity ${row.name} (invalid JSON):`, err);
    }
  }
  return entities;
}

async function loadRelations(pool: Queryable): Promise<Relation[]> {
  const result = await pool.query(
    "SELECT name, definition FROM _relations ORDER BY name",
  );
  const relations: Relation[] = [];
  for (const row of result.rows) {
    try {
      const def =
        typeof row.definition === "string"
          ? JSON.parse(row.definition)
          : row.definition;
      relations.push(def as Relation);
    } catch (err) {
      console.warn(
        `WARN: skipping relation ${row.name} (invalid JSON):`,
        err,
      );
    }
  }
  return relations;
}

async function loadRules(pool: Queryable): Promise<Rule[]> {
  const result = await pool.query(
    "SELECT id, entity, hook, type, definition, priority, active FROM _rules ORDER BY entity, priority",
  );
  const rules: Rule[] = [];
  for (const row of result.rows) {
    try {
      const def =
        typeof row.definition === "string"
          ? JSON.parse(row.definition)
          : row.definition;
      rules.push({
        id: row.id,
        entity: row.entity,
        hook: row.hook,
        type: row.type,
        definition: def as RuleDefinition,
        priority: row.priority,
        active: row.active,
      });
    } catch (err) {
      console.warn(`WARN: skipping rule ${row.id} (invalid JSON):`, err);
    }
  }
  return rules;
}

async function loadStateMachines(pool: Queryable): Promise<StateMachine[]> {
  const result = await pool.query(
    "SELECT id, entity, field, definition, active FROM _state_machines ORDER BY entity",
  );
  const machines: StateMachine[] = [];
  for (const row of result.rows) {
    try {
      const def =
        typeof row.definition === "string"
          ? JSON.parse(row.definition)
          : row.definition;
      machines.push({
        id: row.id,
        entity: row.entity,
        field: row.field,
        definition: normalizeDefinition(def as StateMachineDefinition),
        active: row.active,
      });
    } catch (err) {
      console.warn(
        `WARN: skipping state machine ${row.id} (invalid JSON):`,
        err,
      );
    }
  }
  return machines;
}

async function loadPermissions(pool: Queryable): Promise<Permission[]> {
  const result = await pool.query(
    "SELECT id, entity, action, roles, conditions FROM _permissions ORDER BY entity, action",
  );
  const permissions: Permission[] = [];
  for (const row of result.rows) {
    try {
      const conditions: PermissionCondition[] =
        typeof row.conditions === "string"
          ? JSON.parse(row.conditions)
          : (row.conditions ?? []);
      permissions.push({
        id: row.id,
        entity: row.entity,
        action: row.action,
        roles: row.roles ?? [],
        conditions,
      });
    } catch (err) {
      console.warn(`WARN: skipping permission ${row.id} (invalid JSON):`, err);
    }
  }
  return permissions;
}

async function loadWorkflows(pool: Queryable): Promise<Workflow[]> {
  const result = await pool.query(
    "SELECT id, name, trigger, context, steps, active FROM _workflows ORDER BY name",
  );
  const workflows: Workflow[] = [];
  for (const row of result.rows) {
    try {
      const trigger: WorkflowTrigger =
        typeof row.trigger === "string"
          ? JSON.parse(row.trigger)
          : row.trigger;
      const context: Record<string, string> =
        typeof row.context === "string"
          ? JSON.parse(row.context)
          : (row.context ?? {});
      const rawSteps: any[] =
        typeof row.steps === "string"
          ? JSON.parse(row.steps)
          : (row.steps ?? []);
      workflows.push({
        id: row.id,
        name: row.name,
        trigger,
        context,
        steps: normalizeWorkflowSteps(rawSteps),
        active: row.active,
      });
    } catch (err) {
      console.warn(`WARN: skipping workflow ${row.name} (invalid JSON):`, err);
    }
  }
  return workflows;
}

async function loadWebhooks(pool: Queryable): Promise<Webhook[]> {
  const result = await pool.query(
    "SELECT id, entity, hook, url, method, headers, condition, async, retry, active FROM _webhooks ORDER BY entity, hook",
  );
  const webhooks: Webhook[] = [];
  for (const row of result.rows) {
    try {
      const headers: Record<string, string> =
        typeof row.headers === "string"
          ? JSON.parse(row.headers)
          : (row.headers ?? {});
      const retry: WebhookRetry =
        typeof row.retry === "string"
          ? JSON.parse(row.retry)
          : (row.retry ?? { max_attempts: 3, backoff: "exponential" });
      webhooks.push({
        id: row.id,
        entity: row.entity,
        hook: row.hook,
        url: row.url,
        method: row.method,
        headers,
        condition: row.condition ?? "",
        async: row.async,
        retry,
        active: row.active,
      });
    } catch (err) {
      console.warn(`WARN: skipping webhook ${row.id} (invalid JSON):`, err);
    }
  }
  return webhooks;
}

File: elixir-phoenix/lib/rocket/metadata/loader.ex
defmodule Rocket.Metadata.Loader do
  @moduledoc "Loads all metadata from database into Registry."

  alias Rocket.Store.Postgres
  alias Rocket.Metadata.{Registry, Entity, Relation}

  require Logger

  def load_all(conn, registry) do
    with {:ok, entities} <- load_entities(conn),
         {:ok, relations} <- load_relations(conn) do
      Registry.load(registry, entities, relations)

      {:ok, rules} = load_rules(conn)
      Registry.load_rules(registry, rules)

      {:ok, machines} = load_state_machines(conn)
      Registry.load_state_machines(registry, machines)

      {:ok, workflows} = load_workflows(conn)
      Registry.load_workflows(registry, workflows)

      {:ok, permissions} = load_permissions(conn)
      Registry.load_permissions(registry, permissions)

      {:ok, webhooks} = load_webhooks(conn)
      Registry.load_webhooks(registry, webhooks)

      Logger.info(
        "Loaded #{length(entities)} entities, #{length(relations)} relations, " <>
          "#{length(rules)} rules, #{length(machines)} state machines, " <>
          "#{length(workflows)} workflows, #{length(permissions)} permissions, " <>
          "#{length(webhooks)} webhooks into registry"
      )

      :ok
    end
  end

  def reload(conn, registry), do: load_all(conn, registry)

  # ── Individual loaders ──

  defp load_entities(conn) do
    case Postgres.query_rows(conn, "SELECT name, table_name, definition FROM _entities ORDER BY name") do
      {:ok, rows} ->
        entities =
          rows
          |> Enum.map(fn r ->
            case r["definition"] do
              def_map when is_map(def_map) ->
                merged = Map.merge(def_map, %{"name" => r["name"], "table" => r["table_name"] || r["name"]})
                Entity.from_map(merged)

              _ ->
                nil
            end
          end)
          |> Enum.reject(&is_nil/1)

        {:ok, entities}

      {:error, err} ->
        {:error, err}
    end
  end

  defp load_relations(conn) do
    case Postgres.query_rows(conn, "SELECT name, source, target, definition FROM _relations ORDER BY name") do
      {:ok, rows} ->
        relations =
          rows
          |> Enum.map(fn r ->
            case r["definition"] do
              def_map when is_map(def_map) ->
                merged = Map.merge(def_map, %{"name" => r["name"], "source" => r["source"], "target" => r["target"]})
                Relation.from_map(merged)

              _ ->
                nil
            end
          end)
          |> Enum.reject(&is_nil/1)

        {:ok, relations}

      {:error, err} ->
        {:error, err}
    end
  end

  defp load_rules(conn) do
    case Postgres.query_rows(
           conn,
           "SELECT id, entity, hook, type, definition, priority, active FROM _rules ORDER BY entity, priority"
         ) do
      {:ok, rows} ->
        rules =
          Enum.map(rows, fn r ->
            %Rocket.Metadata.Rule{
              id: r["id"],
              entity: r["entity"],
              hook: r["hook"],
              type: r["type"],
              definition: parse_rule_definition(r["definition"]),
              priority: r["priority"] || 0,
              active: r["active"] != false
            }
          end)

        {:ok, rules}

      {:error, err} ->
        {:error, err}
    end
  end

  defp load_state_machines(conn) do
    case Postgres.query_rows(
           conn,
           "SELECT id, entity, field, definition, active FROM _state_machines ORDER BY entity"
         ) do
      {:ok, rows} ->
        machines =
          Enum.map(rows, fn r ->
            %Rocket.Metadata.StateMachine{
              id: r["id"],
              entity: r["entity"],
              field: r["field"],
              definition: parse_state_machine_definition(r["definition"]),
              active: r["active"] != false
            }
          end)

        {:ok, machines}

      {:error, err} ->
        {:error, err}
    end
  end

  defp load_workflows(conn) do
    case Postgres.query_rows(
           conn,
           "SELECT id, name, trigger, context, steps, active FROM _workflows ORDER BY name"
         ) do
      {:ok, rows} ->
        workflows =
          Enum.map(rows, fn r ->
            %Rocket.Metadata.Workflow{
              id: r["id"],
              name: r["name"],
              trigger: parse_workflow_trigger(r["trigger"]),
              context: r["context"] || %{},
              steps: parse_workflow_steps(r["steps"]),
              active: r["active"] != false
            }
          end)

        {:ok, workflows}

      {:error, err} ->
        {:error, err}
    end
  end

  defp load_permissions(conn) do
    case Postgres.query_rows(
           conn,
           "SELECT id, entity, action, roles, conditions FROM _permissions ORDER BY entity, action"
         ) do
      {:ok, rows} ->
        permissions =
          Enum.map(rows, fn r ->
            %Rocket.Metadata.Permission{
              id: r["id"],
              entity: r["entity"],
              action: r["action"],
              roles: r["roles"] || [],
              conditions: parse_permission_conditions(r["conditions"])
            }
          end)

        {:ok, permissions}

      {:error, err} ->
        {:error, err}
    end
  end

  defp load_webhooks(conn) do
    case Postgres.query_rows(
           conn,
           "SELECT id, entity, hook, url, method, headers, condition, async, retry, active FROM _webhooks ORDER BY entity, hook"
         ) do
      {:ok, rows} ->
        webhooks =
          Enum.map(rows, fn r ->
            %Rocket.Metadata.Webhook{
              id: r["id"],
              entity: r["entity"],
              hook: r["hook"],
              url: r["url"],
              method: r["method"] || "POST",
              headers: r["headers"] || %{},
              condition: r["condition"] || "",
              async: r["async"] != false,
              retry: parse_webhook_retry(r["retry"]),
              active: r["active"] != false
            }
          end)

        {:ok, webhooks}

      {:error, err} ->
        {:error, err}
    end
  end

  # ── Parsers ──

  defp parse_rule_definition(nil), do: %Rocket.Metadata.RuleDefinition{}

  defp parse_rule_definition(map) when is_map(map) do
    %Rocket.Metadata.RuleDefinition{
      field: map["field"],
      operator: map["operator"],
      value: map["value"],
      expression: map["expression"],
      message: map["message"],
      stop_on_fail: map["stop_on_fail"] || false
    }
  end

  defp parse_rule_definition(_), do: %Rocket.Metadata.RuleDefinition{}

  defp parse_state_machine_definition(nil), do: %Rocket.Metadata.StateMachineDefinition{}

  defp parse_state_machine_definition(map) when is_map(map) do
    %Rocket.Metadata.StateMachineDefinition{
      states: map["states"] || [],
      initial_state: map["initial_state"],
      transitions:
        (map["transitions"] || [])
        |> Enum.map(&parse_transition/1)
    }
  end

  defp parse_state_machine_definition(_), do: %Rocket.Metadata.StateMachineDefinition{}

  defp parse_transition(map) when is_map(map) do
    from =
      case map["from"] do
        list when is_list(list) -> list
        str when is_binary(str) -> [str]
        _ -> []
      end

    %Rocket.Metadata.Transition{
      from: from,
      to: map["to"],
      roles: map["roles"] || [],
      guard: map["guard"],
      actions: (map["actions"] || []) |> Enum.map(&parse_transition_action/1)
    }
  end

  defp parse_transition(_), do: nil

  defp parse_transition_action(map) when is_map(map) do
    %Rocket.Metadata.TransitionAction{
      type: map["type"],
      field: map["field"],
      value: map["value"],
      url: map["url"],
      method: map["method"],
      entity: map["entity"],
      data: map["data"]
    }
  end

  defp parse_transition_action(_), do: nil

  defp parse_workflow_trigger(nil), do: %Rocket.Metadata.WorkflowTrigger{}

  defp parse_workflow_trigger(map) when is_map(map) do
    %Rocket.Metadata.WorkflowTrigger{
      type: map["type"],
      entity: map["entity"],
      field: map["field"],
      to: map["to"]
    }
  end

  defp parse_workflow_trigger(_), do: %Rocket.Metadata.WorkflowTrigger{}

  defp parse_workflow_steps(nil), do: []

  defp parse_workflow_steps(list) when is_list(list) do
    Enum.map(list, &parse_workflow_step/1)
  end

  defp parse_workflow_steps(_), do: []

  defp parse_workflow_step(map) when is_map(map) do
    %Rocket.Metadata.WorkflowStep{
      id: map["id"],
      name: map["name"],
      type: map["type"],
      actions: map["actions"] || [],
      condition: map["condition"],
      on_true: parse_step_goto(map["on_true"]),
      on_false: parse_step_goto(map["on_false"]),
      on_approve: parse_step_goto(map["on_approve"]),
      on_reject: parse_step_goto(map["on_reject"]),
      on_timeout: parse_step_goto(map["on_timeout"]),
      deadline: map["deadline"],
      roles: map["roles"] || [],
      next: parse_step_goto(map["next"])
    }
  end

  defp parse_workflow_step(_), do: nil

  defp parse_step_goto(nil), do: nil
  defp parse_step_goto("end"), do: "end"
  defp parse_step_goto(%{"goto" => step_id}), do: step_id
  defp parse_step_goto(other) when is_binary(other), do: other
  defp parse_step_goto(_), do: nil

  defp parse_permission_conditions(nil), do: []

  defp parse_permission_conditions(list) when is_list(list) do
    Enum.map(list, fn c ->
      %Rocket.Metadata.PermissionCondition{
        field: c["field"],
        operator: c["operator"] || "eq",
        value: c["value"]
      }
    end)
  end

  defp parse_permission_conditions(_), do: []

  defp parse_webhook_retry(nil), do: %Rocket.Metadata.WebhookRetry{}

  defp parse_webhook_retry(map) when is_map(map) do
    %Rocket.Metadata.WebhookRetry{
      max_attempts: map["max_attempts"] || 3,
      backoff: map["backoff"] || "exponential"
    }
  end

  defp parse_webhook_retry(_), do: %Rocket.Metadata.WebhookRetry{}
end
